source/server/drain_manager_impl.cc

```diff
@@ -122,7 +140,22 @@ void DrainManagerImpl::startDrainSequence(std::function<void()> drain_complete_c
   }
 
   ASSERT(!drain_tick_timer_);
-  draining_ = true;
+  const std::chrono::seconds drain_delay(server_.options().drainTime());
+
+  // Note https://github.com/envoyproxy/envoy/issues/31457, previous to which,
+  // drain_deadline_ was set *after* draining_ resulting in a read/write race between
+  // the main thread running this function from admin, and the worker thread calling
+  // drainClose. Note that drain_deadline_ is default-constructed which guarantees
+  // to set the time-since epoch to a count of 0
+  // (https://en.cppreference.com/w/cpp/chrono/time_point/time_point).
+  ASSERT(drain_deadline_.time_since_epoch().count() == 0, "drain_deadline_ cannot be set twice.");
+
+  // Since draining_ is atomic, it is safe to set drain_deadline_ without a mutex
+  // as drain_close() only reads from drain_deadline_ if draining_ is true, and
+  // C++ will not re-order an assign to an atomic. See
+  // https://stackoverflow.com/questions/40320254/reordering-atomic-operations-in-c .
```

Comment:
This is only true if set with SeqCst (or maybe release, though I'm hazy on all of this), right? I assume this is the default for assignment using an atomic in C++ but maybe you want to make SeqCst explicit to be clear about it?

Reply:
I believe this is the default based on https://en.cppreference.com/w/cpp/atomic/atomic/store but maybe it'd be clearer if I set the atomic bool by explicitly specifying that with a store call rather than using operator=.

Reply:
done